drivers: gpio: Fix two bugs in nrfx gpio

pdunaj opened the pull request:
Commit fixes two issues:
- Return correct value from gpiote_channel_alloc
- Clear interupt event when callback is disabled

Jira:DESK-270

Fixes #9535

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj committed:
drivers: gpio: Fix two bugs in nrfx gpio

Commit fixes two issues:
- Return correct value from gpiote_channel_alloc
- Check if pending event interrupt was not disabled

Jira:DESK-270

Fixes #9535

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj commented:
Hi @carlescufi , can you review and merge? Also pull it into our fork?

pdunaj commented:
I am not checking rest of the code for correctness. I fixed only two issues that block our application development.

Tested on PCA20041 (zephyr was built using NCS fork). Before the change is applied asserts were triggered, after the fix everything works as expected.

carlescufi commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -55,7 +55,7 @@ static int gpiote_channel_alloc(u32_t abs_pin, nrf_gpiote_polarity_t polarity)
 			nrf_gpiote_event_configure(channel, abs_pin, polarity);
 			nrf_gpiote_event_enable(channel);
 			nrf_gpiote_int_enable(BIT(channel));
-			return channel;
```
Why is this an issue if one checks for return value `>=0` ?

pdunaj replied:
Return from `gpiote_pin_int_cfg` is used directly by `gpio_pin_enable_callback` and `gpio_pin_disable_callback`.

Brief from `gpio_pin_enable_callback` API:

>>* @return 0 if successful, negative errno code on failure.

Application was not expecting a positive value from this call. I can fix the application but the problem is that in many places code checks are `if (err)`. It would be huge problem if underlying layer start to return the positive values without mentioning them in the API.

ioannisg replied:
Not sure I follow completely the explanation from @pdunaj , however, I like this change. This is aligned with several similar cases in Zephyr, where 0 indicates success and non-zero errors map to error codes, reflecting error conditions.

ioannisg replied:
A follow-up question; how does the upper layers get to know the  exact channel that was allocated?

pdunaj replied:
@ioannisg does it need to know? From the point of view of the application we are enabling/disabling callback on a pin. How channel refers to it?
If it is important for any reason maybe there should be a specific API created?

ioannisg replied:
Np, I  was just wondering how and if the channel info is needed by the  upper layer.

Mierunski replied:
It is not needed now, it can stay this way.

carlescufi commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -67,10 +67,14 @@ static void gpiote_channel_free(u32_t abs_pin)
 		    && (intenset & BIT(i))) {
 			nrf_gpiote_event_disable(i);
 			nrf_gpiote_int_disable(BIT(i));
+			nrf_gpiote_event_clear(evt);
```
@Mierunski @anangl can you take another look at this?

pdunaj replied:
both are off today, @pizi-nordic had a look

ioannisg replied:
I am not 100% sure here.  AFAIK, the interrupt  (or some other poll-based-) handler, which processes the corresponding event, is supposed to clear the EVENTS register.

Alternatively, the EVENTS register is cleared when the event is enabled (to make sure we do not have trash value there). 

So this requires some justification, IMO

pdunaj replied:
>> I am not 100% sure here. AFAIK, the interrupt (or some other poll-based-) handler, which processes the corresponding event, is supposed to clear the EVENTS register.

The event is cleared in isr handler. The problem is that when application disables interrupt, no disable interrupt should be received after call is finished. It is easy to imagine the problem if we think about irq_disable or spinlock. Application would not be able to function correctly if late interrupts hit it after call disabling the interrupt is done.

pdunaj replied:
>> Alternatively, the EVENTS register is cleared when the event is enabled (to make sure we do not have trash value there).

I agree this is a problem but I don't want to focus on it as it does not impact our application. I see potential issue when two submodules would use GPIO pins and actively enable/disable callbacks. The way this driver is implemented will cause interrupts to be constantly interfering with each other.

There is also a bigger bug than that. I don't recall any lock taken before on gpiote_pin_int_cfg is called. There is a chance that when two threads would enable callback on two different pins, scheduling race would assign them to the same channel.

ioannisg replied:
If the "late" interrupt hits before line 77, I suppose you have the exact same problem you are trying to solve with line 77, right?

ioannisg replied:
@pdunaj it would be nice if you could elaborate in detail, which problem exactly line 77 is solving. It is not clear from the discussion, here. A step-by-step explanation would be appreciated.

So, I am trying to make few points here, fo which I would appreciate your feedback. 
1. This function de-registers the GPIOTE channel and clears the  hardware interrupt source, associated with this channel. After the interrupt is cleared, i.e. the call in line 76 has returned, it should be the responsibility of the driver and/or the application _not_ to inspect the EVENTS register of a channel (in "thread" mode) we have just de-registered. 

2. If the driver inspects the EVENTS register in IRQ Handler mode, then, as you say, there may be race conditions, so line 77 does not solve anything, IMHO.

3. Typically, in other Nordic drivers in Zephyr, the EVENTS registers are cleared right before an peripheral event is activated, and I would very much like all drivers to comply to this rule. If you think that this is bad/incomplete design, I would of course like to listen :)

pdunaj replied:
@ioannisg , I am not sure if I can say it clearer. The goal is to not receive the interrupt A after interrupt A was disabled.
In the current state this driver does not provide this functionality.
After patch is applied interrupts are disabled and after call to disable is done no callback of this type is received.

Note that I am not the author of this nor I intend to fix other issues I see in this code. I have other stuff to do but until this code is broken I cannot progress. If you think this can or should be done in a better way please give exact code. I would much appreciate it.

pizi-nordic replied:
@pdunaj: Could you please move `rf_gpiote_event_clear(evt)` before `nrf_gpiote_event_enable(channel)` in gpiote_channel_alloc() and check if this solves your issue? If this will not work, we will dig a bit deeper into this driver.

Selected as best solution during discussion with @ioannisg.

pdunaj replied:
I will check it, thanks

pdunaj replied:
I have talked with @pizi-nordic off-line. For anybody that follows this I will explain what is the concern being discussed.

All agree that no interrupt should be received after callback is disabled. @ioannisg expects that even if event is not cleared interrupt should not be triggered after it was disabled by `nrf_gpiote_int_disable`. From this perspective interrupt should be tiggered after being re-enabled by `nrf_gpiote_int_enable` (and just before that event should be cleared as this is the way it is always done).

pdunaj replied:
Unfortunately this is not what I observe in the driver behavior. Interrupt is hit after I disable callbacks and before I re-enable them.
I did the change proposed above by @pizi-nordic and it changed nothing. Below is the log after the change.

`bp x` is displayed in callback (interrupt received from driver). `1` on entry, `2` after callbacks are disabled, `3` at exit.
`disable n` is added close to `int_disable` in `gpiote_channel_free` - n is channel number
`enable n` is added close to `int_enable` in `gpiote_channel_alloc` - n is channel number

As is visible in logs interrupts are disabled but I am being hit by one.

```
e: module_state_event module:main state:ready
enable 0
enable 1
enable 2
enable 3
e: module_state_event module:board state:ready
e: module_state_event module:wheel state:ready
e: module_state_event module:buttons state:ready
[bt] [INF] hci_vs_init: HW Platform: Nordic Semiconductor (0x0002)
[bt] [INF] hci_vs_init: HW Variant: nRF52x (0x0002)
[bt] [INF] hci_vs_init: Firmware: Standard Bluetooth controller (0x00) Version 1.12 Build 99
[bt] [WRN] bt_init: No ID address. Expecting one to come from storage.
[bt] [INF] bt_dev_show_info: Identity: ca:74:81:0b:25:1f (random)
[bt] [INF] bt_dev_show_info: HCI: version 5.0 (0x09) revision 0x0000, manufacturer 0x05f1
[bt] [INF] bt_dev_show_info: LMP: version 5.0 (0x09) subver 0xffff
e: module_state_event module:ble_state state:ready
e: module_state_event module:bas state:ready
e: module_state_event module:dis state:ready
e: module_state_event module:hids state:ready
e: module_state_event module:ble_adv state:ready




bp 1
dissable 0
dissable 1
dissable 2
dissable 3
bp 2
bp 3
bp 1
bp 2
ASSERTION FAIL [!atomic_get(&scanning)] @ /home/padu/work/fw-nrfconnect-nrf/samples/nrf_desktop/src/hw_interface/buttons_pca20041.c:184:
```

ioannisg reviewed (changes requested):
Left some requests for clarifications

carlescufi commented:
@ioannisg can you take another look please?

kl-cruz commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -67,10 +67,14 @@ static void gpiote_channel_free(u32_t abs_pin)
 
 	for (size_t i = 0; i < GPIOTE_CH_NUM; i++) {
+		nrf_gpiote_events_t evt =
+			offsetof(NRF_GPIOTE_Type, EVENTS_IN[i]);
```
It will work correctly, but I do not know if it is not too much. How about changing `offsetof` to `NRF_GPIOTE_EVENTS_IN_0 + i`?

pdunaj replied:
this is copy/paste from isr handler code in the same file

pdunaj commented:
Hi @ioannisg , @carlescufi , @pizi-nordic ,
I understand that more investigation may be needed. At the moment however the driver is broken and should be fixed.

I propose either:
- leave the driver, accept the fix, start investigation and provide subsequent fix or explanation, or
- revert the changes, start investigation and commit again once it works.

carlescufi commented:
@pdunaj I vote for the first option

pdunaj commented:
@carlescufi , me too.

pdunaj commented:
@ioannisg , @carlescufi , @pizi-nordic 
I did some short investigation.

First, our application is configuring the interrupts in edge/double edge mode. @Mierunski , were edge triggered interrupts tested?
I checked how it works in the level mode. In this mode it apparently works apart from one issue. If you press buttons on two pins and disable all pin callbacks in the first interrupt, loop in nrfx gpio isr will still issue second callback (this is a software bug - code should check that interrupt was disabled). Apart from that level mode works with no problems.

As for the edge mode. In documentation it is mentioned that when interrupts are disabled in the interrupt we code must wait a moment before interrupt will be cleared. I think we have enough code to assure that but I added a busy wait there anyway. This does not help.
So it seems that for this mode clearing event after interrupt disable is the only thing that helps. I suspect this is a SoC feature. Maybe it would be worth checking it on some other type...

There is however another issue that I see - SoC cannot leave system off state. The GPIO change that caused the problem that is fixed also changed things in regards with other places where GPIOTE calls were used. One of that was wake up event from off state. It is not working now.
@Mierunski , was waking up from off state tested?

The good thing of all this is that I also found two bugs in the application. Non of them however is related to the issue.

pdunaj commented:
Problem is found and understood thanks to everybody involved.

When interrupts handler is triggered event is cleared. Before interrupts on this channel is disabled another event is received. This received interrupt sets NVIC in pending state. Interrupt will so be triggered regardless of what is the state of events.

The problem is caused by issue in the interrupt handler. To check if callback should be fired handler is only checking if event is set but it not checking that interrupt was disabled.

The first patch that contained clear after disable helped as interrupt handler has no longer seen the event as set. Callback was not fired.

The new patch is also checking that interrupt is enabled and only fires callback if event is set and event interrupt is enabled.

pdunaj commented:
@ioannisg , can you check the fix? If you are happy then please merge after approval.

Mierunski reviewed (approved):
Lets merge these fixes now, I will address other issues in another PR

pdunaj commented:
Hi @ioannisg , can you merge it now?
I forgot that you have merging rights :) This is a very useful information as there is nobody in the office that can merge :)
